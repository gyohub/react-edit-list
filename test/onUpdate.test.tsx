import React from 'react';
import {cleanup, fireEvent, render, waitFor} from '@testing-library/react';

import ReactEditList, * as REL from 'react-edit-list';

import {schema, data, onLoad} from './data';

describe('onUpdate()', () => {
    it('onUpdate() undefined', async () => {
        let recvItem: REL.Row = {};
        let oldItem: REL.Row = {};
        const onUpdate = jest.fn((item: REL.Row, old: REL.Row): void => {
            recvItem = item;
            oldItem = old;
        });
        const r = render(<ReactEditList schema={schema} onLoad={onLoad} onUpdate={onUpdate} />);
        await waitFor(() => expect(r.getByText(/Desk/)).toBeInTheDocument());
        expect(r.container.innerHTML).toMatchSnapshot();

        fireEvent(r.getByText(/Desk/), new MouseEvent('click', {bubbles: true}));
        await waitFor(() => expect(r.container.querySelectorAll('input').length).toBe(3));
        expect(r.container.innerHTML).toMatchSnapshot();

        const inputs = r.container.querySelectorAll('input');
        fireEvent.change(inputs[0], {target: {value: 'new Desk'}});
        fireEvent(await r.findByText('✓'), new MouseEvent('click', {bubbles: true}));

        await waitFor(() => expect(recvItem.product).toBe('new Desk'));
        expect(onUpdate).toBeCalledTimes(1);
        expect(recvItem.type).toBe('1');
        expect(recvItem.price).toBe(100);
        expect(recvItem.stock).toBe(20);
        expect(oldItem.product).toBe('Desk');
        await waitFor(() => expect(r.container.querySelectorAll('input')).toHaveLength(0));
        expect(r.container.innerHTML).toMatchSnapshot();
        r.unmount();
    });

    it('onUpdate() false', async () => {
        let recvItem: REL.Row = {};
        let oldItem: REL.Row = {};
        const onUpdate = jest.fn((item: REL.Row, old: REL.Row): false => {
            recvItem = item;
            oldItem = old;
            return false;
        });
        const r = render(<ReactEditList schema={schema} onLoad={onLoad} onUpdate={onUpdate} />);
        await waitFor(() => expect(r.getByText(/Desk/)).toBeInTheDocument());
        expect(r.container.innerHTML).toMatchSnapshot();

        fireEvent(r.getByText(/Desk/), new MouseEvent('click', {bubbles: true}));
        await waitFor(() => expect(r.container.querySelectorAll('input').length).toBe(3));
        expect(r.container.innerHTML).toMatchSnapshot();

        const inputs = r.container.querySelectorAll('input');
        fireEvent.change(inputs[0], {target: {value: 'Desk 2'}});
        fireEvent(await r.findByText('✓'), new MouseEvent('click', {bubbles: true}));

        await waitFor(() => expect(recvItem.product).toBe('Desk 2'));
        expect(onUpdate).toBeCalledTimes(1);
        expect(recvItem.type).toBe('1');
        expect(recvItem.price).toBe(100);
        expect(recvItem.stock).toBe(20);
        expect(oldItem.product).toBe('Desk');
        await waitFor(() => expect(r.container.querySelectorAll('input')).toHaveLength(3));
        expect(r.container.innerHTML).toMatchSnapshot();
        r.unmount();
    });

    it('onUpdate() replace', async () => {
        let recvItem: REL.Row = {};
        const onUpdate = jest.fn((item: REL.Row): Promise<REL.Row> => {
            recvItem = item;
            const modified = {...item};
            modified.product = 'different';
            return Promise.resolve(modified);
        });
        const r = render(<ReactEditList schema={schema} onLoad={onLoad} onUpdate={onUpdate} />);
        await waitFor(() => expect(r.getByText(/Desk/)).toBeInTheDocument());
        expect(r.container.innerHTML).toMatchSnapshot();

        fireEvent(r.getByText(/Desk/), new MouseEvent('click', {bubbles: true}));
        await waitFor(() => expect(r.container.querySelectorAll('input').length).toBe(3));
        expect(r.container.innerHTML).toMatchSnapshot();

        const inputs = r.container.querySelectorAll('input');
        fireEvent.change(inputs[0], {target: {value: 'test item 3'}});
        fireEvent(await r.findByText('✓'), new MouseEvent('click', {bubbles: true}));

        await waitFor(() => expect(recvItem.product).toBe('test item 3'));
        expect(onUpdate).toBeCalledTimes(1);
        expect(recvItem.type).toBe('1');
        expect(recvItem.price).toBe(100);
        expect(recvItem.stock).toBe(20);
        await waitFor(() => expect(r.container.querySelectorAll('input')).toHaveLength(0));
        await waitFor(() => expect(r.getByText(/different/)));
        expect(r.container.innerHTML).toMatchSnapshot();
        r.unmount();
    });

    it('onUpdate() w/enum w/custom editor', async () => {
        let recvItem: REL.Row = {};
        let oldItem: REL.Row = {};
        const onUpdate = jest.fn((item: REL.Row, old: REL.Row): void => {
            recvItem = item;
            oldItem = old;
        });
        const r = render(
            <ReactEditList
                edit={{
                    product: (props) => <div className='customEditor'>{props.value as string}</div>
                }}
                editProps={{price: {max: 2000}}}
                schema={schema}
                onLoad={onLoad}
                onUpdate={onUpdate}
            />
        );
        await waitFor(() => expect(r.getByText(/Desk/)).toBeInTheDocument());
        expect(r.container.innerHTML).toMatchSnapshot();

        fireEvent(r.getByText(/Desk/), new MouseEvent('click', {bubbles: true}));
        await waitFor(() => expect(r.container.querySelectorAll('input').length).toBe(2));
        expect(r.container.innerHTML).toMatchSnapshot();

        const selects = r.container.querySelectorAll('select');
        fireEvent.change(selects[0], {target: {value: '2'}});
        fireEvent(await r.findByText('✓'), new MouseEvent('click', {bubbles: true}));

        await waitFor(() => expect(recvItem.product).toBe('Desk'));
        expect(onUpdate).toBeCalledTimes(1);
        expect(recvItem.type).toBe('2');
        expect(recvItem.price).toBe(100);
        expect(recvItem.stock).toBe(20);
        expect(oldItem.product).toBe('Desk');
        await waitFor(() => expect(r.container.querySelectorAll('input')).toHaveLength(0));
        expect(r.container.innerHTML).toMatchSnapshot();
        r.unmount();
    });

    it('onUpdate() w/keyboard', async () => {
        let recvItem: REL.Row = {};
        let oldItem: REL.Row = {};
        const onUpdate = jest.fn((item: REL.Row, old: REL.Row): void => {
            recvItem = item;
            oldItem = old;
        });
        const r = render(<ReactEditList schema={schema} onLoad={onLoad} onUpdate={onUpdate} />);
        await waitFor(() => expect(r.getByText(/Desk/)).toBeInTheDocument());
        expect(r.container.innerHTML).toMatchSnapshot();

        fireEvent(r.getByText(/Desk/), new MouseEvent('click', {bubbles: true}));
        await waitFor(() => expect(r.container.querySelectorAll('input').length).toBe(3));
        expect(r.container.innerHTML).toMatchSnapshot();

        const inputs = r.container.querySelectorAll('input');
        fireEvent.change(inputs[1], {target: {value: '102'}});
        fireEvent.keyDown(inputs[1], {key: 'Enter', keyCode: 13, code: 13, charCode: 13});

        await waitFor(() => expect(recvItem.product).toBe('Desk'));
        expect(onUpdate).toBeCalledTimes(1);
        expect(recvItem.type).toBe('1');
        expect(recvItem.price).toBe(102);
        expect(recvItem.stock).toBe(20);
        expect(oldItem.product).toBe('Desk');
        await waitFor(() => expect(r.container.querySelectorAll('input')).toHaveLength(0));
        expect(r.container.innerHTML).toMatchSnapshot();
        r.unmount();
    });

    it('onUpdate() cancel', async () => {
        const onUpdate = jest.fn((item: REL.Row, old: REL.Row): void => undefined);
        const r = render(<ReactEditList schema={schema} onLoad={onLoad} onUpdate={onUpdate} />);
        await waitFor(() => expect(r.getByText(/Desk/)).toBeInTheDocument());
        expect(r.container.innerHTML).toMatchSnapshot();

        fireEvent(r.getByText(/Desk/), new MouseEvent('click', {bubbles: true}));
        await waitFor(() => expect(r.container.querySelectorAll('input').length).toBe(3));
        expect(r.container.innerHTML).toMatchSnapshot();

        const inputs = r.container.querySelectorAll('input');
        fireEvent.change(inputs[2], {target: {value: '102'}});
        fireEvent(await r.getAllByText('x')[0], new MouseEvent('click', {bubbles: true}));
        await waitFor(() => expect(r.container.querySelectorAll('input')).toHaveLength(0));

        expect(onUpdate).toBeCalledTimes(0);
        expect(r.container.innerHTML).toMatchSnapshot();
        r.unmount();
    });

    it('onUpdate() cancel w/keyboard', async () => {
        const onUpdate = jest.fn((item: REL.Row, old: REL.Row): void => undefined);
        const r = render(<ReactEditList schema={schema} onLoad={onLoad} onUpdate={onUpdate} />);
        await waitFor(() => expect(r.getByText(/Desk/)).toBeInTheDocument());
        expect(r.container.innerHTML).toMatchSnapshot();

        fireEvent(r.getByText(/Desk/), new MouseEvent('click', {bubbles: true}));
        await waitFor(() => expect(r.container.querySelectorAll('input').length).toBe(3));
        expect(r.container.innerHTML).toMatchSnapshot();

        const inputs = r.container.querySelectorAll('input');
        fireEvent.change(inputs[0], {target: {value: 'new Desk'}});
        fireEvent.keyDown(inputs[0], {key: 'Escape', keyCode: 27, code: 27, charCode: 27});
        await waitFor(() => expect(r.container.querySelectorAll('input')).toHaveLength(0));

        expect(onUpdate).toBeCalledTimes(0);
        expect(r.container.innerHTML).toMatchSnapshot();
        r.unmount();
    });

    it('disableUpdate', async () => {
        const r = render(<ReactEditList schema={schema} onLoad={onLoad} disableUpdate={true} />);
        await waitFor(() => expect(r.getByText(/Desk/)).toBeInTheDocument());
        expect(r.container.innerHTML).toMatchSnapshot();

        fireEvent(r.getByText(/Desk/), new MouseEvent('click', {bubbles: true}));
        await waitFor(() => expect(r.container.querySelectorAll('input').length).toBe(0));
        expect(r.container.innerHTML).toMatchSnapshot();
        r.unmount();
    });
});
