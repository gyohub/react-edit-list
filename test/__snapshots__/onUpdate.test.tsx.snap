// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`onUpdate() onUpdate() cancel 1`] = `"<table><thead><tr><th>product</th><th>type</th><th>price</th><th>stock</th><th></th></tr></thead><tbody><tr><td>Desk</td><td>capex</td><td>100</td><td>20</td><td><button>x</button></td></tr><tr><td>Printer</td><td>capex</td><td>500</td><td>10</td><td><button>x</button></td></tr><tr><td>Paper</td><td>consumable</td><td>5</td><td>2000</td><td><button>x</button></td></tr><tr><td>Chair</td><td>capex</td><td>50</td><td>50</td><td><button>x</button></td></tr><tr><td>Computer</td><td>capex</td><td>1000</td><td>20</td><td><button>x</button></td></tr><tr><td>Rent</td><td></td><td>2000</td><td></td><td><button>x</button></td></tr><tr><td></td><td></td><td></td><td></td><td>&nbsp;</td></tr></tbody></table>"`;

exports[`onUpdate() onUpdate() cancel 2`] = `"<table><thead><tr><th>product</th><th>type</th><th>price</th><th>stock</th><th></th></tr></thead><tbody><tr><td><input type=\\"text\\" value=\\"Desk\\"></td><td><select><option></option><option value=\\"1\\">capex</option><option value=\\"2\\">consumable</option></select></td><td><input type=\\"number\\" value=\\"100\\"></td><td><input type=\\"number\\" value=\\"20\\"></td><td><button>✓</button><button>x</button></td></tr><tr><td>Printer</td><td>capex</td><td>500</td><td>10</td><td><button>x</button></td></tr><tr><td>Paper</td><td>consumable</td><td>5</td><td>2000</td><td><button>x</button></td></tr><tr><td>Chair</td><td>capex</td><td>50</td><td>50</td><td><button>x</button></td></tr><tr><td>Computer</td><td>capex</td><td>1000</td><td>20</td><td><button>x</button></td></tr><tr><td>Rent</td><td></td><td>2000</td><td></td><td><button>x</button></td></tr><tr><td></td><td></td><td></td><td></td><td>&nbsp;</td></tr></tbody></table>"`;

exports[`onUpdate() onUpdate() cancel 3`] = `"<table><thead><tr><th>product</th><th>type</th><th>price</th><th>stock</th><th></th></tr></thead><tbody><tr><td>Desk</td><td>capex</td><td>100</td><td>20</td><td><button>x</button></td></tr><tr><td>Printer</td><td>capex</td><td>500</td><td>10</td><td><button>x</button></td></tr><tr><td>Paper</td><td>consumable</td><td>5</td><td>2000</td><td><button>x</button></td></tr><tr><td>Chair</td><td>capex</td><td>50</td><td>50</td><td><button>x</button></td></tr><tr><td>Computer</td><td>capex</td><td>1000</td><td>20</td><td><button>x</button></td></tr><tr><td>Rent</td><td></td><td>2000</td><td></td><td><button>x</button></td></tr><tr><td></td><td></td><td></td><td></td><td>&nbsp;</td></tr></tbody></table>"`;

exports[`onUpdate() onUpdate() cancel w/keyboard 1`] = `"<table><thead><tr><th>product</th><th>type</th><th>price</th><th>stock</th><th></th></tr></thead><tbody><tr><td>Desk</td><td>capex</td><td>100</td><td>20</td><td><button>x</button></td></tr><tr><td>Printer</td><td>capex</td><td>500</td><td>10</td><td><button>x</button></td></tr><tr><td>Paper</td><td>consumable</td><td>5</td><td>2000</td><td><button>x</button></td></tr><tr><td>Chair</td><td>capex</td><td>50</td><td>50</td><td><button>x</button></td></tr><tr><td>Computer</td><td>capex</td><td>1000</td><td>20</td><td><button>x</button></td></tr><tr><td>Rent</td><td></td><td>2000</td><td></td><td><button>x</button></td></tr><tr><td></td><td></td><td></td><td></td><td>&nbsp;</td></tr></tbody></table>"`;

exports[`onUpdate() onUpdate() cancel w/keyboard 2`] = `"<table><thead><tr><th>product</th><th>type</th><th>price</th><th>stock</th><th></th></tr></thead><tbody><tr><td><input type=\\"text\\" value=\\"Desk\\"></td><td><select><option></option><option value=\\"1\\">capex</option><option value=\\"2\\">consumable</option></select></td><td><input type=\\"number\\" value=\\"100\\"></td><td><input type=\\"number\\" value=\\"20\\"></td><td><button>✓</button><button>x</button></td></tr><tr><td>Printer</td><td>capex</td><td>500</td><td>10</td><td><button>x</button></td></tr><tr><td>Paper</td><td>consumable</td><td>5</td><td>2000</td><td><button>x</button></td></tr><tr><td>Chair</td><td>capex</td><td>50</td><td>50</td><td><button>x</button></td></tr><tr><td>Computer</td><td>capex</td><td>1000</td><td>20</td><td><button>x</button></td></tr><tr><td>Rent</td><td></td><td>2000</td><td></td><td><button>x</button></td></tr><tr><td></td><td></td><td></td><td></td><td>&nbsp;</td></tr></tbody></table>"`;

exports[`onUpdate() onUpdate() cancel w/keyboard 3`] = `"<table><thead><tr><th>product</th><th>type</th><th>price</th><th>stock</th><th></th></tr></thead><tbody><tr><td>Desk</td><td>capex</td><td>100</td><td>20</td><td><button>x</button></td></tr><tr><td>Printer</td><td>capex</td><td>500</td><td>10</td><td><button>x</button></td></tr><tr><td>Paper</td><td>consumable</td><td>5</td><td>2000</td><td><button>x</button></td></tr><tr><td>Chair</td><td>capex</td><td>50</td><td>50</td><td><button>x</button></td></tr><tr><td>Computer</td><td>capex</td><td>1000</td><td>20</td><td><button>x</button></td></tr><tr><td>Rent</td><td></td><td>2000</td><td></td><td><button>x</button></td></tr><tr><td></td><td></td><td></td><td></td><td>&nbsp;</td></tr></tbody></table>"`;

exports[`onUpdate() onUpdate() false 1`] = `"<table><thead><tr><th>product</th><th>type</th><th>price</th><th>stock</th><th></th></tr></thead><tbody><tr><td>Desk</td><td>capex</td><td>100</td><td>20</td><td><button>x</button></td></tr><tr><td>Printer</td><td>capex</td><td>500</td><td>10</td><td><button>x</button></td></tr><tr><td>Paper</td><td>consumable</td><td>5</td><td>2000</td><td><button>x</button></td></tr><tr><td>Chair</td><td>capex</td><td>50</td><td>50</td><td><button>x</button></td></tr><tr><td>Computer</td><td>capex</td><td>1000</td><td>20</td><td><button>x</button></td></tr><tr><td>Rent</td><td></td><td>2000</td><td></td><td><button>x</button></td></tr><tr><td></td><td></td><td></td><td></td><td>&nbsp;</td></tr></tbody></table>"`;

exports[`onUpdate() onUpdate() false 2`] = `"<table><thead><tr><th>product</th><th>type</th><th>price</th><th>stock</th><th></th></tr></thead><tbody><tr><td><input type=\\"text\\" value=\\"Desk\\"></td><td><select><option></option><option value=\\"1\\">capex</option><option value=\\"2\\">consumable</option></select></td><td><input type=\\"number\\" value=\\"100\\"></td><td><input type=\\"number\\" value=\\"20\\"></td><td><button>✓</button><button>x</button></td></tr><tr><td>Printer</td><td>capex</td><td>500</td><td>10</td><td><button>x</button></td></tr><tr><td>Paper</td><td>consumable</td><td>5</td><td>2000</td><td><button>x</button></td></tr><tr><td>Chair</td><td>capex</td><td>50</td><td>50</td><td><button>x</button></td></tr><tr><td>Computer</td><td>capex</td><td>1000</td><td>20</td><td><button>x</button></td></tr><tr><td>Rent</td><td></td><td>2000</td><td></td><td><button>x</button></td></tr><tr><td></td><td></td><td></td><td></td><td>&nbsp;</td></tr></tbody></table>"`;

exports[`onUpdate() onUpdate() false 3`] = `"<table><thead><tr><th>product</th><th>type</th><th>price</th><th>stock</th><th></th></tr></thead><tbody><tr><td><input type=\\"text\\" value=\\"Desk 2\\"></td><td><select><option></option><option value=\\"1\\">capex</option><option value=\\"2\\">consumable</option></select></td><td><input type=\\"number\\" value=\\"100\\"></td><td><input type=\\"number\\" value=\\"20\\"></td><td><button>✓</button><button>x</button></td></tr><tr><td>Printer</td><td>capex</td><td>500</td><td>10</td><td><button>x</button></td></tr><tr><td>Paper</td><td>consumable</td><td>5</td><td>2000</td><td><button>x</button></td></tr><tr><td>Chair</td><td>capex</td><td>50</td><td>50</td><td><button>x</button></td></tr><tr><td>Computer</td><td>capex</td><td>1000</td><td>20</td><td><button>x</button></td></tr><tr><td>Rent</td><td></td><td>2000</td><td></td><td><button>x</button></td></tr><tr><td></td><td></td><td></td><td></td><td>&nbsp;</td></tr></tbody></table>"`;

exports[`onUpdate() onUpdate() replace 1`] = `"<table><thead><tr><th>product</th><th>type</th><th>price</th><th>stock</th><th></th></tr></thead><tbody><tr><td>Desk</td><td>capex</td><td>100</td><td>20</td><td><button>x</button></td></tr><tr><td>Printer</td><td>capex</td><td>500</td><td>10</td><td><button>x</button></td></tr><tr><td>Paper</td><td>consumable</td><td>5</td><td>2000</td><td><button>x</button></td></tr><tr><td>Chair</td><td>capex</td><td>50</td><td>50</td><td><button>x</button></td></tr><tr><td>Computer</td><td>capex</td><td>1000</td><td>20</td><td><button>x</button></td></tr><tr><td>Rent</td><td></td><td>2000</td><td></td><td><button>x</button></td></tr><tr><td></td><td></td><td></td><td></td><td>&nbsp;</td></tr></tbody></table>"`;

exports[`onUpdate() onUpdate() replace 2`] = `"<table><thead><tr><th>product</th><th>type</th><th>price</th><th>stock</th><th></th></tr></thead><tbody><tr><td><input type=\\"text\\" value=\\"Desk\\"></td><td><select><option></option><option value=\\"1\\">capex</option><option value=\\"2\\">consumable</option></select></td><td><input type=\\"number\\" value=\\"100\\"></td><td><input type=\\"number\\" value=\\"20\\"></td><td><button>✓</button><button>x</button></td></tr><tr><td>Printer</td><td>capex</td><td>500</td><td>10</td><td><button>x</button></td></tr><tr><td>Paper</td><td>consumable</td><td>5</td><td>2000</td><td><button>x</button></td></tr><tr><td>Chair</td><td>capex</td><td>50</td><td>50</td><td><button>x</button></td></tr><tr><td>Computer</td><td>capex</td><td>1000</td><td>20</td><td><button>x</button></td></tr><tr><td>Rent</td><td></td><td>2000</td><td></td><td><button>x</button></td></tr><tr><td></td><td></td><td></td><td></td><td>&nbsp;</td></tr></tbody></table>"`;

exports[`onUpdate() onUpdate() replace 3`] = `"<table><thead><tr><th>product</th><th>type</th><th>price</th><th>stock</th><th></th></tr></thead><tbody><tr><td>different</td><td>capex</td><td>100</td><td>20</td><td><button>x</button></td></tr><tr><td>Printer</td><td>capex</td><td>500</td><td>10</td><td><button>x</button></td></tr><tr><td>Paper</td><td>consumable</td><td>5</td><td>2000</td><td><button>x</button></td></tr><tr><td>Chair</td><td>capex</td><td>50</td><td>50</td><td><button>x</button></td></tr><tr><td>Computer</td><td>capex</td><td>1000</td><td>20</td><td><button>x</button></td></tr><tr><td>Rent</td><td></td><td>2000</td><td></td><td><button>x</button></td></tr><tr><td></td><td></td><td></td><td></td><td>&nbsp;</td></tr></tbody></table>"`;

exports[`onUpdate() onUpdate() undefined 1`] = `"<table><thead><tr><th>product</th><th>type</th><th>price</th><th>stock</th><th></th></tr></thead><tbody><tr><td>Desk</td><td>capex</td><td>100</td><td>20</td><td><button>x</button></td></tr><tr><td>Printer</td><td>capex</td><td>500</td><td>10</td><td><button>x</button></td></tr><tr><td>Paper</td><td>consumable</td><td>5</td><td>2000</td><td><button>x</button></td></tr><tr><td>Chair</td><td>capex</td><td>50</td><td>50</td><td><button>x</button></td></tr><tr><td>Computer</td><td>capex</td><td>1000</td><td>20</td><td><button>x</button></td></tr><tr><td>Rent</td><td></td><td>2000</td><td></td><td><button>x</button></td></tr><tr><td></td><td></td><td></td><td></td><td>&nbsp;</td></tr></tbody></table>"`;

exports[`onUpdate() onUpdate() undefined 2`] = `"<table><thead><tr><th>product</th><th>type</th><th>price</th><th>stock</th><th></th></tr></thead><tbody><tr><td><input type=\\"text\\" value=\\"Desk\\"></td><td><select><option></option><option value=\\"1\\">capex</option><option value=\\"2\\">consumable</option></select></td><td><input type=\\"number\\" value=\\"100\\"></td><td><input type=\\"number\\" value=\\"20\\"></td><td><button>✓</button><button>x</button></td></tr><tr><td>Printer</td><td>capex</td><td>500</td><td>10</td><td><button>x</button></td></tr><tr><td>Paper</td><td>consumable</td><td>5</td><td>2000</td><td><button>x</button></td></tr><tr><td>Chair</td><td>capex</td><td>50</td><td>50</td><td><button>x</button></td></tr><tr><td>Computer</td><td>capex</td><td>1000</td><td>20</td><td><button>x</button></td></tr><tr><td>Rent</td><td></td><td>2000</td><td></td><td><button>x</button></td></tr><tr><td></td><td></td><td></td><td></td><td>&nbsp;</td></tr></tbody></table>"`;

exports[`onUpdate() onUpdate() undefined 3`] = `"<table><thead><tr><th>product</th><th>type</th><th>price</th><th>stock</th><th></th></tr></thead><tbody><tr><td>new Desk</td><td>capex</td><td>100</td><td>20</td><td><button>x</button></td></tr><tr><td>Printer</td><td>capex</td><td>500</td><td>10</td><td><button>x</button></td></tr><tr><td>Paper</td><td>consumable</td><td>5</td><td>2000</td><td><button>x</button></td></tr><tr><td>Chair</td><td>capex</td><td>50</td><td>50</td><td><button>x</button></td></tr><tr><td>Computer</td><td>capex</td><td>1000</td><td>20</td><td><button>x</button></td></tr><tr><td>Rent</td><td></td><td>2000</td><td></td><td><button>x</button></td></tr><tr><td></td><td></td><td></td><td></td><td>&nbsp;</td></tr></tbody></table>"`;

exports[`onUpdate() onUpdate() w/enum 1`] = `"<table><thead><tr><th>product</th><th>type</th><th>price</th><th>stock</th><th></th></tr></thead><tbody><tr><td>Desk</td><td>capex</td><td>100</td><td>20</td><td><button>x</button></td></tr><tr><td>Printer</td><td>capex</td><td>500</td><td>10</td><td><button>x</button></td></tr><tr><td>Paper</td><td>consumable</td><td>5</td><td>2000</td><td><button>x</button></td></tr><tr><td>Chair</td><td>capex</td><td>50</td><td>50</td><td><button>x</button></td></tr><tr><td>Computer</td><td>capex</td><td>1000</td><td>20</td><td><button>x</button></td></tr><tr><td>Rent</td><td></td><td>2000</td><td></td><td><button>x</button></td></tr><tr><td></td><td></td><td></td><td></td><td>&nbsp;</td></tr></tbody></table>"`;

exports[`onUpdate() onUpdate() w/enum 2`] = `"<table><thead><tr><th>product</th><th>type</th><th>price</th><th>stock</th><th></th></tr></thead><tbody><tr><td><input type=\\"text\\" value=\\"Desk\\"></td><td><select><option></option><option value=\\"1\\">capex</option><option value=\\"2\\">consumable</option></select></td><td><input type=\\"number\\" value=\\"100\\"></td><td><input type=\\"number\\" value=\\"20\\"></td><td><button>✓</button><button>x</button></td></tr><tr><td>Printer</td><td>capex</td><td>500</td><td>10</td><td><button>x</button></td></tr><tr><td>Paper</td><td>consumable</td><td>5</td><td>2000</td><td><button>x</button></td></tr><tr><td>Chair</td><td>capex</td><td>50</td><td>50</td><td><button>x</button></td></tr><tr><td>Computer</td><td>capex</td><td>1000</td><td>20</td><td><button>x</button></td></tr><tr><td>Rent</td><td></td><td>2000</td><td></td><td><button>x</button></td></tr><tr><td></td><td></td><td></td><td></td><td>&nbsp;</td></tr></tbody></table>"`;

exports[`onUpdate() onUpdate() w/enum 3`] = `"<table><thead><tr><th>product</th><th>type</th><th>price</th><th>stock</th><th></th></tr></thead><tbody><tr><td>Desk</td><td>consumable</td><td>100</td><td>20</td><td><button>x</button></td></tr><tr><td>Printer</td><td>capex</td><td>500</td><td>10</td><td><button>x</button></td></tr><tr><td>Paper</td><td>consumable</td><td>5</td><td>2000</td><td><button>x</button></td></tr><tr><td>Chair</td><td>capex</td><td>50</td><td>50</td><td><button>x</button></td></tr><tr><td>Computer</td><td>capex</td><td>1000</td><td>20</td><td><button>x</button></td></tr><tr><td>Rent</td><td></td><td>2000</td><td></td><td><button>x</button></td></tr><tr><td></td><td></td><td></td><td></td><td>&nbsp;</td></tr></tbody></table>"`;

exports[`onUpdate() onUpdate() w/keyboard 1`] = `"<table><thead><tr><th>product</th><th>type</th><th>price</th><th>stock</th><th></th></tr></thead><tbody><tr><td>Desk</td><td>capex</td><td>100</td><td>20</td><td><button>x</button></td></tr><tr><td>Printer</td><td>capex</td><td>500</td><td>10</td><td><button>x</button></td></tr><tr><td>Paper</td><td>consumable</td><td>5</td><td>2000</td><td><button>x</button></td></tr><tr><td>Chair</td><td>capex</td><td>50</td><td>50</td><td><button>x</button></td></tr><tr><td>Computer</td><td>capex</td><td>1000</td><td>20</td><td><button>x</button></td></tr><tr><td>Rent</td><td></td><td>2000</td><td></td><td><button>x</button></td></tr><tr><td></td><td></td><td></td><td></td><td>&nbsp;</td></tr></tbody></table>"`;

exports[`onUpdate() onUpdate() w/keyboard 2`] = `"<table><thead><tr><th>product</th><th>type</th><th>price</th><th>stock</th><th></th></tr></thead><tbody><tr><td><input type=\\"text\\" value=\\"Desk\\"></td><td><select><option></option><option value=\\"1\\">capex</option><option value=\\"2\\">consumable</option></select></td><td><input type=\\"number\\" value=\\"100\\"></td><td><input type=\\"number\\" value=\\"20\\"></td><td><button>✓</button><button>x</button></td></tr><tr><td>Printer</td><td>capex</td><td>500</td><td>10</td><td><button>x</button></td></tr><tr><td>Paper</td><td>consumable</td><td>5</td><td>2000</td><td><button>x</button></td></tr><tr><td>Chair</td><td>capex</td><td>50</td><td>50</td><td><button>x</button></td></tr><tr><td>Computer</td><td>capex</td><td>1000</td><td>20</td><td><button>x</button></td></tr><tr><td>Rent</td><td></td><td>2000</td><td></td><td><button>x</button></td></tr><tr><td></td><td></td><td></td><td></td><td>&nbsp;</td></tr></tbody></table>"`;

exports[`onUpdate() onUpdate() w/keyboard 3`] = `"<table><thead><tr><th>product</th><th>type</th><th>price</th><th>stock</th><th></th></tr></thead><tbody><tr><td>Desk</td><td>capex</td><td>102</td><td>20</td><td><button>x</button></td></tr><tr><td>Printer</td><td>capex</td><td>500</td><td>10</td><td><button>x</button></td></tr><tr><td>Paper</td><td>consumable</td><td>5</td><td>2000</td><td><button>x</button></td></tr><tr><td>Chair</td><td>capex</td><td>50</td><td>50</td><td><button>x</button></td></tr><tr><td>Computer</td><td>capex</td><td>1000</td><td>20</td><td><button>x</button></td></tr><tr><td>Rent</td><td></td><td>2000</td><td></td><td><button>x</button></td></tr><tr><td></td><td></td><td></td><td></td><td>&nbsp;</td></tr></tbody></table>"`;
